test/test_torch.py

```diff
@@ -17093,27 +17093,137 @@ def test_rdiv(self, device, dtype):
+    def test_fmod_by_zero_integral(self, device, dtype):
+        # check integral tensor fmod to zero
+        x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+        zero = torch.zeros_like(x)
+        # out
+        out = torch.empty(0, device=device, dtype=dtype)
+        # In-place
+        x_ = x.clone()
+        # RuntimeError on CPU
+        if device == 'cpu':
+            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                x.fmod(zero)
+            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                torch.fmod(x, zero, out=out)
+            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                x.fmod_(zero)
+        # Different value for different dtype on GPU
+        else:
+            if dtype == torch.int64:
+                self.assertEqual(x.fmod(zero) == 4294967295, x >= 0)
+                self.assertEqual(x.fmod(zero) == -1, x < 0)
+                # out
+                torch.fmod(x, zero, out=out)
+                self.assertEqual(out == 4294967295, x >= 0)
+                self.assertEqual(out == -1, x < 0)
+                self.assertEqual(out.size(), torch.Size([10, 10]))
+                # in-place
+                x_.fmod_(zero)
+                self.assertEqual(x_ == 4294967295, x >= 0)
+                self.assertEqual(x_ == -1, x < 0)
+            else:
+                value = 255 if dtype == torch.uint8 else -1
+                self.assertTrue(torch.all(x.fmod(zero) == value))
+                # out
+                torch.fmod(x, zero, out=out)
+                self.assertTrue(torch.all(out == value))
+                self.assertEqual(out.size(), torch.Size([10, 10]))
+                # in-place
+                x_.fmod_(zero)
+                self.assertTrue(torch.all(x_ == value))
+
+    # TO DO
+    # Cross product dtypes after fix type promotion
+    # Issue #47779: https://github.com/pytorch/pytorch/issues/47779
     @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False, include_bool=False, include_complex=False))
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
     def test_fmod(self, device, dtype):
-        m1 = torch.Tensor(10, 10).uniform_(-10., 10.).to(dtype=dtype, device=device)
-        res1 = m1.clone()
-        q = 3
-        res1[:, 3].fmod_(q)
-        res2 = m1.clone()
-        for i in range(m1.size(1)):
-            res2[i, 3] = math.fmod(res2[i, 3], q)
-        self.assertEqual(res1, res2)
+        # Use numpy as reference
+        def _reference_implementation(x, mod):
+            res = torch.fmod(x, mod)
+            out = torch.empty(0, device=device, dtype=dtype)
+            torch.fmod(x, mod, out=out)
+            x_ = x.clone()
+            # Integral Tensor in-place fmod raises RuntimeError due to invalid type casting
+            # It will be solved by type promotion #47779
+            if device != 'xla':
+                x_.fmod_(mod)
```

Comment:
Integral Tensor doing in-place fmod raises invalid type-casting Issue. It can be solved as the input should be promoted to floating-point.

Reply:
Change to `if device in ['cpu', 'cuda']:` to get rid of XLA test.